packages/components/src/tabs/tablist.tsx

```diff
@@ -17,19 +23,106 @@ import type { TabListProps } from './types';
 import { useTabsContext } from './context';
 import { TabListWrapper } from './styles';
 import type { WordPressComponentProps } from '../context';
+import type { CSSProperties } from 'react';
+
+function useTrackElementOffset(
+	targetElement?: HTMLElement | null,
+	onUpdate?: () => void
+) {
+	const [ indicatorPosition, setIndicatorPosition ] = useState( {
+		left: 0,
+		top: 0,
+		width: 0,
+		height: 0,
+	} );
+
+	// TODO: replace with useEventCallback or similar when officially available.
+	const updateCallbackRef = useRef( onUpdate );
+	useLayoutEffect( () => {
+		updateCallbackRef.current = onUpdate;
+	} );
+
+	const observedElementRef = useRef< HTMLElement >();
+	const resizeObserverRef = useRef< ResizeObserver >();
+	useEffect( () => {
+		if ( targetElement === observedElementRef.current ) return;
+		observedElementRef.current = targetElement ?? undefined;
+
+		function updateIndicator( element: HTMLElement ) {
+			setIndicatorPosition( {
+				left: element.offsetLeft,
+				top: element.offsetTop,
+				width: element.offsetWidth,
+				height: element.offsetHeight,
+			} );
+			updateCallbackRef.current?.();
+		}
+
+		// Set up a ResizeObserver.
+		if ( ! resizeObserverRef.current ) {
+			resizeObserverRef.current = new ResizeObserver( () => {
+				if ( observedElementRef.current )
+					updateIndicator( observedElementRef.current );
+			} );
+		}
+		const { current: resizeObserver } = resizeObserverRef;
+
+		// Unobserve previous element.
+		const { current: observedElement } = observedElementRef;
+		if ( observedElement ) resizeObserver.unobserve( observedElement );
+
+		// Observe new element.
+		if ( targetElement ) {
+			updateIndicator( targetElement );
+			resizeObserver.observe( targetElement );
+		}
+	}, [ targetElement ] );
+
+	return indicatorPosition;
+}
+
+type ValueUpdateContext = {
+	previousValue: unknown;
+};
+
+function useOnValueUpdate(
+	value: unknown,
+	onUpdate: ( context: ValueUpdateContext ) => void
+) {
+	const [ previousValue, setPreviousValue ] = useState( value );
+
+	useEffect( () => {
+		if ( previousValue !== value ) {
+			onUpdate( { previousValue } );
+			setPreviousValue( value );
+		}
+	}, [ previousValue, value, onUpdate ] );
+}
```

Comment:
It needs to only trigger on updates (not initially), and knowledge of the previous value is necessary to bail on some updates too. The useEffect dep array is not enough. This is a common hook util I've needed sooner or later on every design system I've built :P

Reply:
The hook is fine, but you'd be better off storing the previous value in a ref:
```js
const previousValue = useRef( value );
useEffect( () => {
  if ( previousValue.current !== value ) {
    onUpdate( { previousValue: previousValue.current } );
    previousValue.current = value;
  }
}, [ value, onUpdate ] );
```
`useState` triggers an unnecessary extra render when `value` changes, but `useRef` doesn't.

The hook could be also more precisely typed:
```tsx
useOnValueUpdate<T>(value: T, onUpdate: (context: ValueUpdateContext<T>) => void)
```

One day, if used widely, it could be moved to `@wordpress/compose`. There is a `usePrevious` hook there which is somewhat related.

Another slight improvement is storing the `onUpdate` callback in a ref, so that the value-updating effect is triggered only by `value` change and nothing else. Note that the effect is a noop anyway when `onUpdate` changes.
```js
const latestOnUpdate = useRef( onUpdate );
useEffect( () => {
  latestOnUpdate.current = onUpdate;
}, [ onUpdate ] );

const previousValue = useRef( value );
useEffect( () => {
  if ( previousValue.current !== value ) {
    latestOnUpdate.current( { previousValue: previousValue.current } );
    previousValue.current = value;
  }
}, [ value ] );
```

Reply:
Thank you for the notes @jsnajdr. I already have these changes locally (not yet pushed as I'm debugging something), you read my mind there :P

While the function ref works like that, I tend to use a layout effect just to make sure it runs as early as possible. It's what is [typically used in `useEventCallback` implementations](https://github.com/facebook/react/issues/14099#issuecomment-440013892) too (and you can also often see it without abstractions, e.g. [see Kent C. Dodds take](https://epicreact.dev/the-latest-ref-pattern-in-react/)). Hopefully we'll just be able to use [something like `useEvent`](https://github.com/reactjs/rfcs/pull/220) if it officially lands in React.

Reply:
Btw, I'm already doing the function ref thing a few lines above: https://github.com/WordPress/gutenberg/pull/60560/files/283f10952c38cf35c184042ed9e11134e2fd1d71#diff-64263a335082a1c717edd91225f49242f93a3b2625646af8af012dc96a6ff7a6R40

Reply:
> Btw, I'm already doing the function ref thing a few lines above:

I think it's enough when the `useOnValueUpdate` hook does it. In the calling component, IMO it's a bit pointless.

Reply:
It's not used in the calling component though :P

Reply:
Oh, disregard my previous comment, I totally misread the code. I thought that it stores the `onUpdate` callback to a ref and then calls `useOnUpdateValue` with the ref's `.current` value.

Reply:
No worries!